Load state dict post hook

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75424
* __->__ #75422

Adds load_state_dict_post_hook. Proposal: https://github.com/pytorch/pytorch/issues/75287

Went with no args except for module optionally for now. Additional args can be
added on use case basis.

A few notes:
- Similar to state_dict post hook, we run it after this module and all submodule children have been loaded.
- Added try/finally to ensure the hook runs even if load_state_dict throws.

Differential Revision: [D35439200](https://our.internmc.facebook.com/intern/diff/D35439200/)

rohan-varma committed:
Load state dict post hook

Adds load_state_dict_post_hook. Proposal: https://github.com/pytorch/pytorch/issues/75287

Went with no args except for module optionally for now. Additional args can be
added on use case basis.

A few notes:
- Similar to state_dict post hook, we run it after this module and all submodule children have been loaded.
- Added try/finally to ensure the hook runs even if load_state_dict throws.

Differential Revision: [D35439200](https://our.internmc.facebook.com/intern/diff/D35439200/)

[ghstack-poisoned]

awgu commented on torch/nn/modules/module.py:
```diff
@@ -1537,12 +1562,18 @@ def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
+                    if child is not None:
+                        load(child, prefix + name + '.')
+            finally:
+                # Note that post hooks for a module are called after self _and_ all
```
nit: duplicate `all`

zhaojuanmao commented on torch/nn/modules/module.py:
```diff
@@ -1407,6 +1410,27 @@ def _register_load_state_dict_pre_hook(self, hook, with_module=False):
         self._load_state_dict_pre_hooks[handle.id] = hook
         return handle
 
+    def _register_load_state_dict_post_hook(self, hook, with_module=False):
```
curious why "with_module" option is needed?

jbschlosser replied:
Looks like this was done to match `_register_load_state_dict_pre_hook()`, where the `with_module` flag was added later on.

albanD replied:
Note that we wrote a full issue on how we want these hooks to happen with Rohan. This PR is outdated wrt that and should be updated to match.

rohan-varma committed:
Update on "Load state dict post hook"


Adds load_state_dict_post_hook. Proposal: https://github.com/pytorch/pytorch/issues/75287

Went with no args except for module optionally for now. Additional args can be
added on use case basis.

A few notes:
- Similar to state_dict post hook, we run it after this module and all submodule children have been loaded.
- Added try/finally to ensure the hook runs even if load_state_dict throws.

Differential Revision: [D35439200](https://our.internmc.facebook.com/intern/diff/D35439200/)

[ghstack-poisoned]

malfet reviewed (changes requested):
Not requesting any changes, just rewoking my review as PR has not been approved internally (nor publically)

rohan-varma commented:
Closing in favor of new API to be implemented as part of addressing https://github.com/pytorch/pytorch/issues/75287
